### PR TITLE
Add tril & triu kernels and op

### PR DESCRIPTION
**Pull request by AlienLiang23**
Add tril & triu kernels and op

### CODE REVIEW

**Comment by AlienLiang23** [comment]
triu passes all UTs 
image

**Comment by fengyuan14** [comment]
Cases look good to me. python_ref covers all dtypes.

File: src/aten/TriangluarOps.cpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
+#include <ATen/ATen.h>
+#include <ATen/XPUNativeFunctions.h>
+#include <ATen/core/Tensor.h>
+#include <ATen/core/op_registration/adaption.h>
+#include <aten/sycl/TriangularOpsKernel.h>
+#include <comm/RegisterUtils.h>
+
+namespace at {
+
+void meta_func_tril(const Tensor& self, int64_t k) {
```
Comment (fengyuan14) [inline src/aten/TriangluarOps.cpp:10]:
```suggestion
void tril_meta(const Tensor& self, int64_t k) {
```

File: src/aten/TriangluarOps.cpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+Tensor& XPUNativeFunctions::tril_out(
+    const Tensor& self,
+    int64_t diagonal,
+    Tensor& out) {
+  std::optional<Device> common_device = std::nullopt;
+  c10::impl::check_and_update_common_device(
+      common_device, out, "xpu::tril_out", "out");
+  c10::impl::check_and_update_common_device(
+      common_device, self, "xpu::tril_out", "self");
+  meta_func_tril(self, diagonal);
```
Comment (fengyuan14) [inline src/aten/TriangluarOps.cpp:24]:
```suggestion
  tril_meta(self, diagonal);
```

File: src/aten/TriangluarOps.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  xpu::resize_out(out, self.sizes(), {}, self.options());
+  return native::xpu::tril_xpu_kernel(out, self, diagonal);
+}
+
+Tensor XPUNativeFunctions::tril(const Tensor& self, int64_t diagonal) {
+  meta_func_tril(self, diagonal);
```
Comment (fengyuan14) [inline src/aten/TriangluarOps.cpp:30]:
```suggestion
  tril_meta(self, diagonal);
```

File: src/aten/TriangluarOps.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  Tensor out = xpu::create_out(self.sizes(), {}, self.options());
+  return tril_out(self, diagonal, out);
+}
+
+Tensor& XPUNativeFunctions::tril_(Tensor& self, int64_t diagonal) {
+  meta_func_tril(self, diagonal);
```
Comment (fengyuan14) [inline src/aten/TriangluarOps.cpp:36]:
```suggestion
  tril_meta(self, diagonal);
```

File: src/aten/TriangluarOps.cpp [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  xpu::check_inplace(self, self.sizes(), self.options());
+  return tril_out(self, diagonal, self);
+}
+
+void meta_func_triu(const Tensor& self, int64_t k) {
```
Comment (fengyuan14) [inline src/aten/TriangluarOps.cpp:41]:
```suggestion
void triu_meta(const Tensor& self, int64_t k) {
```

File: src/aten/TriangluarOps.cpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+Tensor& XPUNativeFunctions::triu_out(
+    const Tensor& self,
+    int64_t diagonal,
+    Tensor& out) {
+  std::optional<Device> common_device = std::nullopt;
+  c10::impl::check_and_update_common_device(
+      common_device, out, "xpu::triu_out", "out");
+  c10::impl::check_and_update_common_device(
+      common_device, self, "xpu::triu_out", "self");
+  meta_func_triu(self, diagonal);
```
Comment (fengyuan14) [inline src/aten/TriangluarOps.cpp:55]:
```suggestion
  triu_meta(self, diagonal);
```

File: src/aten/TriangluarOps.cpp [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  xpu::resize_out(out, self.sizes(), {}, self.options());
+  return native::xpu::triu_xpu_kernel(out, self, diagonal);
+}
+
+Tensor XPUNativeFunctions::triu(const Tensor& self, int64_t diagonal) {
+  meta_func_triu(self, diagonal);
```
Comment (fengyuan14) [inline src/aten/TriangluarOps.cpp:61]:
```suggestion
  triu_meta(self, diagonal);
```

File: src/aten/TriangluarOps.cpp [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  Tensor out = xpu::create_out(self.sizes(), {}, self.options());
+  return triu_out(self, diagonal, out);
+}
+
+Tensor& XPUNativeFunctions::triu_(Tensor& self, int64_t diagonal) {
+  meta_func_triu(self, diagonal);
```
Comment (fengyuan14) [inline src/aten/TriangluarOps.cpp:67]:
```suggestion
  triu_meta(self, diagonal);
```

File: src/aten/sycl/TriangularOpsKernel.cpp [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+#define TRIU_TRIL_LAMBDA(upper)                                   \
+  [&] {                                                           \
+    if (canUse32BitIndexMath(self)) {                             \
+      apply_triu_tril<scalar_t, int32_t, upper>(result, self, k); \
+    } else {                                                      \
+      apply_triu_tril<scalar_t, int64_t, upper>(result, self, k); \
+    }                                                             \
+  }
+
+Tensor& tril_xpu_kernel(Tensor& result, const Tensor& self, int64_t k) {
```
Comment (fengyuan14) [inline src/aten/sycl/TriangularOpsKernel.cpp:121]:
```suggestion
Tensor& tril_kernel(Tensor& result, const Tensor& self, int64_t k) {
```

File: src/aten/sycl/TriangularOpsKernel.h [hunk 10]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+#include <ATen/native/TensorIterator.h>
+
+namespace at::native::xpu {
+
+Tensor& tril_xpu_kernel(Tensor& result, const Tensor& self, int64_t k);
```
Comment (fengyuan14) [inline src/aten/sycl/TriangularOpsKernel.h:6]:
```suggestion
Tensor& tril_kernel(Tensor& result, const Tensor& self, int64_t k);
```

File: src/aten/sycl/TriangularOpsKernel.h [hunk 11]
```diff
@@ -0,0 +1,10 @@
...
+
+Tensor& triu_xpu_kernel(Tensor& result, const Tensor& self, int64_t k);
```
Comment (fengyuan14) [inline src/aten/sycl/TriangularOpsKernel.h:8]:
```suggestion
Tensor& triu_kernel(Tensor& result, const Tensor& self, int64_t k);
```

File: src/aten/sycl/TriangularOpsKernel.cpp [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+      at::ScalarType::BFloat16,
+      at::ScalarType::ComplexHalf,
+      self.scalar_type(),
+      "tril_xpu",
+      TRIU_TRIL_LAMBDA(false));
+
+  return result;
+}
+
+Tensor& triu_xpu_kernel(Tensor& result, const Tensor& self, int64_t k) {
```
Comment (fengyuan14) [inline src/aten/sycl/TriangularOpsKernel.cpp:141]:
```suggestion
Tensor& triu_kernel(Tensor& result, const Tensor& self, int64_t k) {
```

File: src/aten/TriangluarOps.cpp [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    int64_t diagonal,
+    Tensor& out) {
+  std::optional<Device> common_device = std::nullopt;
+  c10::impl::check_and_update_common_device(
+      common_device, out, "xpu::tril_out", "out");
+  c10::impl::check_and_update_common_device(
+      common_device, self, "xpu::tril_out", "self");
+  meta_func_tril(self, diagonal);
+  xpu::resize_out(out, self.sizes(), {}, self.options());
+  return native::xpu::tril_xpu_kernel(out, self, diagonal);
```
Comment (fengyuan14) [inline src/aten/TriangluarOps.cpp:26]:
```suggestion
  return native::xpu::tril_kernel(out, self, diagonal);
```

File: src/aten/TriangluarOps.cpp [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    int64_t diagonal,
+    Tensor& out) {
+  std::optional<Device> common_device = std::nullopt;
+  c10::impl::check_and_update_common_device(
+      common_device, out, "xpu::triu_out", "out");
+  c10::impl::check_and_update_common_device(
+      common_device, self, "xpu::triu_out", "self");
+  meta_func_triu(self, diagonal);
+  xpu::resize_out(out, self.sizes(), {}, self.options());
+  return native::xpu::triu_xpu_kernel(out, self, diagonal);
```
Comment (fengyuan14) [inline src/aten/TriangluarOps.cpp:57]:
```suggestion
  return native::xpu::triu_kernel(out, self, diagonal);
```

File: src/aten/sycl/TriangularOpsKernels.cpp [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+#define TRIU_TRIL_LAMBDA(upper)                                   \
+  [&] {                                                           \
+    if (canUse32BitIndexMath(self)) {                             \
+      apply_triu_tril<scalar_t, int32_t, upper>(result, self, k); \
+    } else {                                                      \
+      apply_triu_tril<scalar_t, int64_t, upper>(result, self, k); \
+    }                                                             \
+  }
+
+Tensor& tril_kernel(Tensor& result, const Tensor& self, int64_t k) {
```
Comment (fengyuan14) [inline src/aten/sycl/TriangularOpsKernels.cpp:121]: Add inplace case handle.
image

File: src/aten/sycl/TriangularOpsKernels.cpp [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+      at::ScalarType::BFloat16,
+      at::ScalarType::ComplexHalf,
+      self.scalar_type(),
+      "tril_xpu",
+      TRIU_TRIL_LAMBDA(false));
+
+  return result;
+}
+
+Tensor& triu_kernel(Tensor& result, const Tensor& self, int64_t k) {
```
Comment (fengyuan14) [inline src/aten/sycl/TriangularOpsKernels.cpp:141]: Ditto

File: src/aten/sycl/TriangularOpsKernels.cpp (hunk 15 above)
Comment (fengyuan14) [reply src/aten/sycl/TriangularOpsKernels.cpp:121]: Reviewed CUDA implementation. Existing CUDA implementation gets a code-path for inplace, but it should be a performance optimization. We will have a separate task for kernel performance optimization. Ported IPEX implementation bases on non-optimized CUDA implementation. but is supposed to support inplace function. Please add an example for inplace function check.

Comment (AlienLiang23) [reply src/aten/sycl/TriangularOpsKernels.cpp:121]: image
inplace function has been covered

**Review by fengyuan14 — APPROVED** [review]

**Pull request merged**